### ISSUE TITLE
[TASK] Define future naming convention for composer scripts

### ISSUE THREAD

**Issue by jpmschuler** [TYPO3-Documentation/tea#450]
Proposal
====
As more and more stuff will be added to composer scripts (#449), there should be a naming convention which allows for easier comprehension and discoverability, especially as additional scripts should allow variants (e.g. running a command with local tools vs with docker, with PHP 7.4 vs PHP 8.1 and so on).
While this is basically the case now, there are opportunities for improvement

Rules
===
Semantics:

1. Task category should be first element in script name
2. `ci` shouldn't be a task category as the scripts should be context agnostic (except if it is a variant)
3. Each script should have a parent script, which either
    - groups them (e.g. `test` groups `test:unit` and `test:functional`) (☑️ done already yet)
    - is the default variant (e.g. `test:functional` pointing to `test:functional:local` as the default, and an alternative `test:functional:docker` exists) (wasn't necessary until now)
    - or if both don't apply, echo a helping text (e.g. it won't make much sense to point `build` to `build:docs`, as that is probably the least expected build artifact=
4. variants/options should be last element
    
Syntax:
1. Aspects like task categories, Levels, Languages, Tasks, Tools need to be separated by colon (:) (☑️ done already yet)
2. Multiple words per part should be avoided, else be in kebab-case (e.g. `coverage:create-directories` instead of  `coverage:createDirectories` or  `coverage:create:directories`) (☑️ done mostly yet)

First Level Categories
====

Thus the suggestion is to have major task categories, currently 

- check (static tests)
- fix (available fixers for static tests)
- test (unit, functional, acceptance)
- build (docs)

Selected Suggestions:
====

This would allow variants like

Suggestion | CurrentCommand
-- | --
build:docs | docs:generate
check | ci:static
check:json | ci:json:lint
check:php:copypaste | ci:php:copypaste
check:php:cs-fixer | ci:php:cs-fixer
check:php:lint | ci:php:lint
check:php:lint:docker |  
check:php:lint:docker:php74 |  
check:php:lint:docker:php80 |  
check:php:lint:docker:php81 |  
check:php:lint:local |  
check:php:sniff | ci:php:sniff
check:php:stan | ci:php:stan
check:php:stan:baseline | phpstan:baseline
check:typoscript:lint | ci:ts:lint
check:yaml:lint | ci:yaml:lint
fix |  
fix:json:composer | ci:composer:normalize
fix:php | fix:php
fix:php:cs | fix:php:cs
fix:php:sniff | fix:php:sniff
test:coverage |  
test:coverage:merge | ci:coverage:merge
test:coverage:prepare | coverage:create-directories
test:functional | ci:tests:functional
test:functional:coverage |  
test:functional:coverage:local | ci:coverage:functional
test:functional:coverage:docker |  
test:functional:docker |  
test:functional:docker:php74 |  
test:functional:docker:php74:mariadb |  
test:functional:docker:php74:postgres |  
test:functional:docker:php80 |  
test:functional:local |  
test:unit | ci:tests:unit
test:unit:coverage | ci:coverage:unit
  | link-extension

PainPoints of the Proposal
===========
a) Preparation stuff
While we have `check:php:sniff` and `fix:php:sniff` as these are separate task categories, I suggest using `check:php:stan:baseline` because I believe tasks like `:init` and `:prepare` should be suffixes in order to ease discoverability, however I don't deem a specific term important, but would rather use `:baseline` here

b) Metrics as a result, rather than a task group
In order to reduce the number of top level task groups I suggest `test:functional:coverage:docker`, however I understand that `metrics:functional` or `metrics:coverage:functional` or `coverage:functional` might be meaningful as well.

c) naming of languages
`tslint` should be called `typoscript:lint` - because I thought it would lint my .ts files (JavaScript TypeScript variant) and it fits that all others are file extensions.
`style` should be called `css` in that regard, (because it would prevent the discussion about styleSheets vs style-sheets vs stylesheets) 

d) awkward commands
`fix:json:composer` sounds ridiculous, but in fact is more selfexplanatory than `composer normalise` and would e.g. allow `fix:json:package` to point to `fixpack` the npm tool which does the same

**Comment by oliverklee** [comment]
Thanks! I like the generall approach a lot and will have a closer look at the proposed list of names and the pain points.

**Comment by oliverklee** [comment]
Some thoughts:

- I like the consistency, and that the concept is easy to follow. :heart:
- We also should have a `check:php` command that runs all PHP-related checks. (This is useful when you're modifying only PHP files.)
- What does `test:coverage` do?
- Maybe we can use the occasion and rename `fix:php:cs` and `fix:php:sniff` to `fix:php:fixer` and `fix:php:sniffer` (as this is closer to the tool names "PHP Coding Standards Fixer" and `PHP_CodeSniffer". (Ditto for the corresponding `check:*` scripts.)
- We should also have a command `check:json:composer` that checks if the `composer.json` is normalized.
- Concerning the `:baseline` suffix: Yes, that sounds good to me.
- I agree with your propose concerning `ts` vs. `typoscript` and `style` vs. `css`.
- I'm also not completely happy  with `fix:json:composer`, but currently cannot come up with a better idea. So I'm okay with using that name and (when/if we come up with something better) changing it later.

**Comment by brotkrueml** [comment]
My random thoughts:

- `check:php:cs-fixer` should be named `check:php:cs`. I would recommend `cs` (or `cgl` to be more in-line with Core) instead of `fixer` (for `check:php:cs` and `fix:php:cs`). The last one is too close to the tool we are using. I think, the command should be tool-agnostic (when it can be). So I want to check the code styles: go with `cs` or `cgl`.
- Instead of `check:json:composer` I propose `check:composer:config`. This is independent from the format, the tool is the relevant player. This way we can provide in the future maybe also `check:composer:valid`.

Overall, this suggestion looks good and is much better structured the the current solution.

**Comment by oliverklee** [comment]
I like `check:composer:config`!

**Comment by oliverklee** [comment]
Concerning `fix:php:cs`: We're actually using two tools here (PHP_CodeSniffer and PHP-CS-Fixer). So we could have three commands (so that we can run the tools independently at least on CI):

- `check:php:cs:fixer`
- `check:php:cs:sniffer`
- `check:php:cs` which then calls the other two tools

@brotkrueml What do you think?

**Comment by brotkrueml** [comment]
Yes, I can live with that ;-) So we are free to update to other tools in the future and leave `check:php:cs` untouched.

**Comment by oliverklee** [comment]
@lukaszuznanski Would you be fine with this?

**Comment by DanielSiepmann** [comment]
> Concerning `fix:php:cs`: We're actually using two tools here (PHP_CodeSniffer and PHP-CS-Fixer). So we could have three commands (so that we can run the tools independently at least on CI):

We use `[symplify/](https://packagist.org/packages/symplify/)easy-coding-standard` which combines both with a single configuration and CLI tool, for PHP only. I don't want to bloat this discussion, just point out that it might be useful to switch to that dependency?

**Comment by d-s-codappix** [comment]
I guess this is obsolete with https://github.com/TYPO3BestPractices/tea/issues/1750
